params/roadmap.py

```diff
@@ -1,72 +1,214 @@
         ...,
-        title='Storage Efficiency (% As fraction)',
-        description='For every KWH entered, how mush is able to be drawn out.'
+        title="Storage Efficiency (%)",
+        description="For every KWH entered, how mush is able to be drawn out.",
     )
     storage_discharge_p: RoadmapParam = Field(
         ...,
-        title='Solar Discharge Depth (% As fraction)',
-        description='How much of the battery\'s capacity can be drawn out at once.'
+        title="Solar Discharge Depth (%)",
+        description="How much of the battery's capacity can be drawn out at once.",
     )
 
-    @validator('end_year')
+    _params: tuple[RoadmapParam, ...]
+
+    @validator("end_year")
     def v_end_year(cls, end_year: float, values: dict[str, int | RoadmapParam]):
-        assert end_year >= values['start_year'], \
-            'end_year must be greater or equal to start_year'
+        assert (
+            end_year >= values["start_year"]
+        ), "end_year must be greater or equal to start_year"
         return end_year
 
+    def __init__(self, **data):
+        super().__init__(**data)
+
+        self._params = (
+            self.solar_gen_kw,
+            self.wind_gen_kw,
+            self.solar_gen_kw,
+            self.storage_efficiency_p,
+            self.storage_discharge_p,
+        )
+
+    def _scenario_from_ends(self, end_values: t.Sequence[int]) -> Scenario:
+        """
+        Create a scenario from the given end values
+        and the years in this Roadmap.
+
+        :param end_values: Value for the end year for each Scenario parameter, ordered.
+        :return: A constructed Scenario.
+        """
+        return Scenario(
+            *(
+                np.linspace(
+                    start=param.start,
+                    stop=end_val,
+                    num=self.end_year - self.start_year,
+                )
+                for param, end_val in zip(self._params, end_values)
+            )
+        )
+
     @property
-    def scenarios(self) -> list[Scenario]:
-        # TODO
-        return []
+    def scenarios(self) -> t.Iterator[Scenario]:
+        """
+        All possible `Scenario` that can be generated form this `Roadmap`.
+        """
+        # a range of end values for each parameter
+        end_value_ranges = (
+            range(param.end_min, param.end_max, param.step) for param in self._params
+        )
+
+        # all combinations of end values
+        ends = product(*end_value_ranges)
+
+        return remove_duplicates(
+            map(
+                self._scenario_from_ends,
+                ends,
+            )
+        )
+
+
+if __name__ == "__main__":
+
+    # mostly copied from google-sheets
+    r = Roadmap(
+        start_year=2020,
+        end_year=2050,
+        solar_gen_kw=RoadmapParam(
+            start=4_000, end_min=50_000, end_max=150_000, step=20_000
+        ),
+        wind_gen_kw=RoadmapParam(start=80, end_min=250, end_max=3_000, step=100),
```

Comment:
complement the earlier suggestion

```suggestion
        solar_capacity_kw=RoadmapParam(
            start=4_000, end_min=50_000, end_max=150_000, step=20_000
        ),
        wind_capacity_kw=RoadmapParam(start=80, end_min=250, end_max=3_000, step=100),
```